src/render/idyll-display/components/author-tool-buttons.js

```diff
@@ -0,0 +1,333 @@
+import React from 'react';
+import Context from '../../context';
+import { DropTarget } from 'react-dnd';
+import { getNodeById, isChildOf, getRandomId } from '../utils/';
+
+const AST = require('idyll-ast');
+const compile = require('idyll-compiler');
+
+const cursorPositions = {};
+
+function getCurrentCursorPosition(parent) {
+  var selection = window.getSelection(),
+    charCount = -1,
+    node;
+
+  if (selection.focusNode) {
+    if (isChildOf(selection.focusNode, parent)) {
+      node = selection.focusNode;
+      charCount = selection.focusOffset;
+
+      while (node) {
+        if (node === parent) {
+          break;
+        }
+
+        if (node.previousSibling) {
+          node = node.previousSibling;
+          charCount += node.textContent.length;
+        } else {
+          node = node.parentNode;
+          if (node === null) {
+            break;
+          }
+        }
+      }
+    }
+  }
+
+  return charCount;
+}
+
+function createRange(node, chars, range) {
+  if (!range) {
+    range = document.createRange();
+    range.selectNode(node);
+    range.setStart(node, 0);
+  }
+
+  if (chars.count === 0) {
+    range.setEnd(node, chars.count);
+  } else if (node && chars.count > 0) {
+    if (node.nodeType === Node.TEXT_NODE) {
+      if (node.textContent.length < chars.count) {
+        chars.count -= node.textContent.length;
+      } else {
+        range.setEnd(node, chars.count);
+        chars.count = 0;
+      }
+    } else {
+      for (var lp = 0; lp < node.childNodes.length; lp++) {
+        range = createRange(node.childNodes[lp], chars, range);
+
+        if (chars.count === 0) {
```

Comment:
Is this check necessary? Will `chars.count` ever be 0 if we're in this `else` branch? I'm currently not seeing how we could be in this branch with the count being 0. It seems like if `chars.count === 0` being true will always make us enter the `if (chars.count === 0)` branch above.

Reply:
Ah so a lot of the logic from author-tool-buttons is from the old AuthorView logic that was there before. I kept it there in case but if it's not needed (currently, it's not being used anywhere I think), I'll delete!

Reply:
O, in that case, I think we can leave as is. And clean it up in another PR!